### DISCUSSION TITLE
ItemInput and Type Example

### DISCUSSION THREAD

**Discussion by francescoemiliani** [forntoh/LcdMenu#43] [Q&A] [answered]
Hello, I'm a new user. And a newbie in arduino. Can someone provide an example for ItemInput?

I played with this library for some days, i have edit it a little bit for initialize the menu without re-initialize lcd (in my project i have a main screen, and the menu is called with a button long press).

I'm doing various tests/expreriment with this library (I still have to study how to change a variable with the menu), but I'm stuck with the ItemInput and the function "Type". I managed to write a single character with this

`menu.type(typing[3]);`

"typing" is an array of number from 0 to 9, but with this code obviously it will type only the third character in the array.

I definitely didn't understand how to use intemInput and Type function. How i can use it properly?

(in this specific case I only want change individually each number of a float variable, but i want understand the use of the library )

Very sorry for my English and for disturb.

**Comment by forntoh** [comment, marked as answer]
The `type()` function is used to enter one character at a time, each character will be appended to the previous character.
I didn't consider every use-case of the `ItemInput`, your question has made me see a new use-case 😊.

Full guide [here](https://lcdmenu.forntoh.dev/examples/input-from-character-set)

**Comment by francescoemiliani** [comment]
If the `type()` function is used to enter one character, i suppose that i correct understand the proper use of `type()` so, there are no need of an example!! (but i appreciate your kindness, thanks!! :smiley: ). This use case is exceptional with a keypad!!!!

I try to create a menu with a very cheap 5 button keyboard, so I'm using some creativity! :smile: 

End of my answer but, to thank you for your kindness, there is a long version:

My skills as a programmer date back to 15 years ago, so so currently I'm not the best. Years ago I written a menu (from scratch) for a PIC controller, when an item is selected by the menu cursor, press enter button, and a  blink cursor was shown, with left and right button you can move the text cursor, and with the up and bottom button, you can change the number highlighted under the text cursor), press enter again and the edited text is confirmed (written in a variable).

I'm trying to reproduce this in arduino, it is highly probable that, also in this case, i will create the entire menu from scratch :yum: . If time and my skills (after 15 years... :roll_eyes: ) allow me to do it (unlikely :sweat_smile: ), i will be happy to communicate how i solved!!

**Reply by forntoh** [reply]
😁 cool,  that's going to be in **v2.1.0,** I implemented the feature last night (the use-case I was talking about 😅).
I am doing some testing before releasing it.

### Summary:
1. Define your charset (list of accepted/valid chars)
1. Move to an ItemInput menu item.
1. Execute `enter()` to go to edit mode. (cursor will start blinking 😉)
1. Use `up()` and `down()` to cycle through your charset.
1. Execute `type(char c)` to write the selected char.
1. Optional: Use `left()` and `right()` to move through already typed characters.
1. Call `back()` to quit edit mode, the callback bound to this item will be executed.

**Comment by wildan32** [comment]
The beautiful code, i am very interested . but has anyone know .
when Iteminput character or value  will saved after enter  , but when arduino restart there is nothing value again. can you show me to load last value again with data from eeprom please give me example

**Reply by forntoh** [reply]
#9 #56
